module/zfs/zio_checksum.c

```diff
@@ -351,7 +351,26 @@ zio_checksum_compute(zio_t *zio, enum zio_checksum checksum,
 		eck->zec_cksum = cksum;
 	} else {
 		ci->ci_func[0](abd, size, spa->spa_cksum_tmpls[checksum],
-		    &bp->blk_cksum);
+		    &cksum);
+
+		if (BP_IS_ENCRYPTED(bp)) {
+			/*
+			 * Weak checksums do not have their entropy spread
+			 * evenly across the bits of the checksum. Therefore,
+			 * when truncating a weak checksum we XOR the first
+			 * 2 words with the last 2 so that we don't "lose" any
+			 * entropy unnecessarily.
```

Comment:
I'm not aware of any, but I know all secure checksums support being truncated without issue whereas there may be weaknesses for XOR folding.